app/src/main/java/io/github/fate_grand_automata/imaging/TesseractOcrService.kt

```diff
@@ -36,29 +50,58 @@ class TesseractOcrService @Inject constructor(
         }
     }
 
-    protected fun finalize() {
-        tessApi.recycle()
-    }
-
     private fun extractTesseractTrainingData() {
         val tessDir = File(context.filesDir.absolutePath, "tessdata")
         if (!tessDir.exists()) {
             tessDir.mkdir()
         }
-        for (assetFileName in context.assets.list("tessdata")!!) {
+        val assetList = context.assets.list("tessdata")
+        if (assetList == null) {
+            Timber.e("Failed to list assets in tessdata")
+            return
+        }
+        for (assetFileName in assetList) {
             val targetFile = File(tessDir, assetFileName)
-            if (!targetFile.exists()) {
```

Comment:
The existing code only copies the file if the training data don't exist.

Did you add this function so the training data can be updated?

Reply:
so that the training data is copied correctly in the off chance that there's some failure